File: Raven.Client.Lightweight/Document/SessionOperations/QueryOperation.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Reflection;
using System.Text.RegularExpressions;
using Raven.Client.Connection;
using Raven.Imports.Newtonsoft.Json.Linq;
using Raven.Abstractions.Data;
using Raven.Abstractions.Linq;
using Raven.Abstractions.Logging;
using Raven.Client.Exceptions;
using Raven.Json.Linq;
using Raven.Abstractions.Extensions;

namespace Raven.Client.Document.SessionOperations
{
    public class QueryOperation
    {
#if !DNXCORE50
        private readonly static ILog log = LogManager.GetCurrentClassLogger();
#else
        private readonly static ILog log = LogManager.GetLogger(typeof(QueryOperation));
#endif

        private readonly InMemoryDocumentSessionOperations sessionOperations;
        private readonly string indexName;
        private readonly IndexQuery indexQuery;
        private readonly bool waitForNonStaleResults;
        private bool disableEntitiesTracking;
        private readonly TimeSpan timeout;
        private readonly Func<IndexQuery, IEnumerable<object>, IEnumerable<object>> transformResults;
        private readonly HashSet<string> includes;
        private QueryResult currentQueryResults;
        private readonly string[] projectionFields;
        private bool firstRequest = true;

        public QueryResult CurrentQueryResults
        {
            get { return currentQueryResults; }
        }

        public string IndexName
        {
            get { return indexName; }
        }

        public IndexQuery IndexQuery
        {
            get { return indexQuery; }
        }

        private Stopwatch sp;

        public QueryOperation(InMemoryDocumentSessionOperations sessionOperations, string indexName, IndexQuery indexQuery,
                              string[] projectionFields, bool waitForNonStaleResults, TimeSpan timeout,
                              Func<IndexQuery, IEnumerable<object>, IEnumerable<object>> transformResults,
                              HashSet<string> includes, bool disableEntitiesTracking)
        {
            this.indexQuery = indexQuery;
            this.waitForNonStaleResults = waitForNonStaleResults;
            this.timeout = timeout;
            this.transformResults = transformResults;
            this.includes = includes;
            this.projectionFields = projectionFields;
            this.sessionOperations = sessionOperations;
            this.indexName = indexName;
            this.disableEntitiesTracking = disableEntitiesTracking;

            AssertNotQueryById();
        }

        private static readonly Regex idOnly = new Regex(@"^__document_id \s* : \s* ([\w_\-/\\\.]+) \s* $",
            RegexOptions.Compiled |
 RegexOptions.IgnorePatternWhitespace);

        private void AssertNotQueryById()
        {
            // this applies to dynamic indexes only
            if (!indexName.StartsWith("dynamic/", StringComparison.OrdinalIgnoreCase) &&
                !string.Equals(indexName, "dynamic", StringComparison.OrdinalIgnoreCase))
                return;

            var match = idOnly.Match(IndexQuery.Query);
            if (match.Success == false)
                return;

            if (sessionOperations.Conventions.AllowQueriesOnId)
                return;

            var value = match.Groups[1].Value.Replace("\\/","/");

            throw new InvalidOperationException("Attempt to query by id only is blocked, you should use call session.Load(\"" + value + "\"); instead of session.Query().Where(x=>x.Id == \"" + value + "\");" + Environment.NewLine + "You can turn this error off by specifying documentStore.Conventions.AllowQueriesOnId = true;, but that is not recommend and provided for backward compatibility reasons only.");
        }

        private void StartTiming()
        {
            sp = Stopwatch.StartNew();
        }

        public void LogQuery()
        {
            if (log.IsDebugEnabled)
            log.Debug("Executing query '{0}' on index '{1}' in '{2}'",
                                          indexQuery.Query, indexName, sessionOperations.StoreIdentifier);
            }

        public IDisposable EnterQueryContext()
        {
            if (firstRequest)
            {
                StartTiming();
                firstRequest = false;
            }

            if (waitForNonStaleResults == false)
                return null;

            return sessionOperations.DocumentStore.DisableAggressiveCaching();
        }

        public bool ShouldQueryAgain(Exception e)
        {
            if (e is NonAuthoritativeInformationException == false)
                return false;

            return sp.Elapsed <= sessionOperations.NonAuthoritativeInformationTimeout;
        }

        public IList<T> Complete<T>()
        {
            var queryResult = currentQueryResults.CreateSnapshot();
            foreach (var include in SerializationHelper.RavenJObjectsToJsonDocuments(queryResult.Includes))
            {
                sessionOperations.TrackIncludedDocument(include);
            }

            var list = queryResult.Results
                .Select(x => x != null ? Deserialize<T>(x) : default(T))
                .ToList();

            if (disableEntitiesTracking == false)
                sessionOperations.RegisterMissingIncludes(queryResult.Results.Where(x => x != null), includes);

            if (transformResults == null)
                return list;

            return transformResults(indexQuery, list.Cast<object>()).Cast<T>().ToList();
        }

        public bool DisableEntitiesTracking
        {
            get { return disableEntitiesTracking; }
            set { disableEntitiesTracking = value; }
        }

        public T Deserialize<T>(RavenJObject result)
        {
            var metadata = result.Value<RavenJObject>("@metadata");
            if ((projectionFields == null || projectionFields.Length <= 0) &&
                (metadata != null && string.IsNullOrEmpty(metadata.Value<string>("@id")) == false))
            {
                return sessionOperations.TrackEntity<T>(metadata.Value<string>("@id"),
                                                        result,
                                                        metadata, disableEntitiesTracking);
            }

            if (typeof(T) == typeof(RavenJObject))
                return (T)(object)result;

            if (typeof(T) == typeof(object) && string.IsNullOrEmpty(result.Value<string>("$type")))
            {
                return (T)(object)new DynamicJsonObject(result);
            }

            var documentId = result.Value<string>(Constants.DocumentIdFieldName); //check if the result contain the reserved name

            if (!string.IsNullOrEmpty(documentId) && typeof(T) == typeof(string) && // __document_id is present, and result type is a string
                // We are projecting one field only (although that could be derived from the
                // previous check, one could never be too careful
                projectionFields != null && projectionFields.Length == 1 && 
                HasSingleValidProperty(result, metadata) // there are no more props in the result object
                )
            {
                return (T)(object)documentId;
            }

            sessionOperations.HandleInternalMetadata(result);

            var deserializedResult = DeserializedResult<T>(result);

            if (string.IsNullOrEmpty(documentId) == false)
            {
                // we need to make an additional check, since it is possible that a value was explicitly stated
                // for the identity property, in which case we don't want to override it.
                var identityProperty = sessionOperations.Conventions.GetIdentityProperty(typeof(T));
                if (identityProperty != null &&
                    (result[identityProperty.Name] == null ||
                     result[identityProperty.Name].Type == JTokenType.Null))
                {
                    sessionOperations.GenerateEntityIdOnTheClient.TrySetIdentity(deserializedResult, documentId);
                }
            }

            return deserializedResult;
        }

        private bool HasSingleValidProperty(RavenJObject result, RavenJObject metadata)
        {
            if (metadata == null && result.Count == 1)
                return true;// { Foo: val }

            if ((metadata != null && result.Count == 2))
                return true; // { @metadata: {}, Foo: val }

            if ((metadata != null && result.Count == 3))
            {
                var entityName = metadata.Value<string>(Constants.RavenEntityName);

                var idPropName = sessionOperations.Conventions.FindIdentityPropertyNameFromEntityName(entityName);

                if (result.ContainsKey(idPropName))
                {
                    // when we try to project the id by name
                    var token = result.Value<RavenJToken>(idPropName);

                    if (token == null || token.Type == JTokenType.Null)
                        return true; // { @metadata: {}, Foo: val, Id: null }
                }
            }

            return false;
        }


        private T DeserializedResult<T>(RavenJObject result)
        {
            var type = typeof(T);
            if (type == typeof(string) || typeof(T).IsValueType() || typeof(T).IsEnum())
            {
                if (projectionFields != null && projectionFields.Length == 1) // we only select a single field
                {
                    return result.Value<T>(projectionFields[0]);
                }
                switch (result.Count)
                {
                    case 1:
                        return result.Value<T>(result.Keys.First());
                    case 2:
                        if(result.ContainsKey(Constants.Metadata))
                            return result.Value<T>(result.Keys.First(x=>x!=Constants.Metadata));
                        break;
                }
            }

            var jsonSerializer = sessionOperations.Conventions.CreateSerializer();
            var ravenJTokenReader = new RavenJTokenReader(result);

            var resultTypeString = result.Value<string>("$type");
            if (string.IsNullOrEmpty(resultTypeString))
            {
                return (T)jsonSerializer.Deserialize(ravenJTokenReader, typeof(T));
            }

            var resultType = Type.GetType(resultTypeString, false);
            if (resultType == null) // couldn't find the type, let us give it our best shot
            {
                return (T)jsonSerializer.Deserialize(ravenJTokenReader, typeof(T));
            }

            return (T)jsonSerializer.Deserialize(ravenJTokenReader, resultType);
        }

        public void ForceResult(QueryResult result)
        {
            currentQueryResults = result;
            currentQueryResults.EnsureSnapshot();
        }

        public bool IsAcceptable(QueryResult result)
        {
            if (sessionOperations.AllowNonAuthoritativeInformation == false &&
                result.NonAuthoritativeInformation)
            {
                if (sp.Elapsed > sessionOperations.NonAuthoritativeInformationTimeout)
                {
                    sp.Stop();
                    throw new TimeoutException(string.Format("Waited for {0:#,#;;0}ms for the query to return authoritative result.", sp.ElapsedMilliseconds));
                }

                if (log.IsDebugEnabled)
                log.Debug(
                        "Non authoritative query results on authoritative query '{0}' on index '{1}' in '{2}', query will be retried, index etag is: {3}",
                            indexQuery.Query,
                            indexName,
                            sessionOperations.StoreIdentifier,
                            result.IndexEtag);
                return false;
            }
            if (waitForNonStaleResults && result.IsStale)
            {
                if (sp.Elapsed > timeout)
                {
                    sp.Stop();

                    throw new TimeoutException(string.Format("Waited for {0:#,#;;0}ms for the query to return non stale result.", sp.ElapsedMilliseconds));
                }

                if (log.IsDebugEnabled)
                log.Debug(
                        "Stale query results on non stale query '{0}' on index '{1}' in '{2}', query will be retried, index etag is: {3}",
                            indexQuery.Query,
                            indexName,
                            sessionOperations.StoreIdentifier,
                            result.IndexEtag);
                return false;
            }
            currentQueryResults = result;
            currentQueryResults.EnsureSnapshot();
            if (log.IsDebugEnabled)
            log.Debug("Query returned {0}/{1} {2}results", result.Results.Count,
                                              result.TotalResults, result.IsStale ? "stale " : "");
            return true;
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/document/sessionoperations/QueryOperation.java
package net.ravendb.client.document.sessionoperations;

import java.io.IOException;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map.Entry;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import net.ravendb.abstractions.basic.SharpEnum;
import net.ravendb.abstractions.basic.Tuple;
import net.ravendb.abstractions.closure.Action2;
import net.ravendb.abstractions.data.Constants;
import net.ravendb.abstractions.data.IndexQuery;
import net.ravendb.abstractions.data.QueryResult;
import net.ravendb.abstractions.extensions.JsonExtensions;
import net.ravendb.abstractions.json.linq.JTokenType;
import net.ravendb.abstractions.json.linq.RavenJArray;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.abstractions.json.linq.RavenJToken;
import net.ravendb.abstractions.json.linq.RavenJValue;
import net.ravendb.abstractions.logging.ILog;
import net.ravendb.abstractions.logging.LogManager;
import net.ravendb.client.document.InMemoryDocumentSessionOperations;
import net.ravendb.client.exceptions.NonAuthoritativeInformationException;
import net.ravendb.client.utils.UrlUtils;

import org.apache.commons.lang.StringUtils;


public class QueryOperation {

  private static final ILog log = LogManager.getCurrentClassLogger();
  private final InMemoryDocumentSessionOperations sessionOperations;
  private final String indexName;
  private final IndexQuery indexQuery;
  private final Set<Tuple<String, Class<?>>> sortByHints;
  private final Action2<String, String> setOperationHeaders;
  private final boolean waitForNonStaleResults;
  private boolean disableEntitiesTracking;
  private final long timeout;
  private final Set<String> includes;
  private QueryResult currentQueryResults;
  private final String[] projectionFields;
  private boolean firstRequest = true;

  private static final Pattern ID_ONLY =  Pattern.compile("^__document_id\\s*:\\s*([\\w_\\-/\\\\\\.]+)\\s*$");

  public QueryResult getCurrentQueryResults() {
    return currentQueryResults;
  }

  public String getIndexName() {
    return indexName;
  }

  public IndexQuery getIndexQuery() {
    return indexQuery;
  }

  private long spStart;
  private long spStop;

  public QueryOperation(InMemoryDocumentSessionOperations sessionOperations, String indexName, IndexQuery indexQuery,
      String[] projectionFields, Set<Tuple<String, Class<?>>> sortByHints,
      boolean waitForNonStaleResults, Action2<String, String> setOperationHeaders, long timeout,
      Set<String> includes, boolean disableEntitiesTracking) {
    this.indexQuery = indexQuery;
    this.sortByHints = sortByHints;
    this.waitForNonStaleResults = waitForNonStaleResults;
    this.setOperationHeaders = setOperationHeaders;
    this.timeout = timeout;
    this.includes = includes;
    this.projectionFields = projectionFields;
    this.sessionOperations = sessionOperations;
    this.indexName = indexName;
    this.disableEntitiesTracking = disableEntitiesTracking;

    assertNotQueryById();
    addOperationHeaders();
  }

  private void assertNotQueryById() {
    // this applies to dynamic indexes only
    if (!indexName.toLowerCase().startsWith("dynamic/") && !StringUtils.equalsIgnoreCase(indexName, "dynamic")) {
      return ;
    }

    Matcher matcher = ID_ONLY.matcher(indexQuery.getQuery());
    if (!matcher.matches()) {
      return ;
    }

    if (sessionOperations.getConventions().isAllowQueriesOnId()) {
      return ;
    }

    String value = matcher.group(1);
    throw new IllegalStateException( "Attempt to query by id only is blocked, you should use call session.load(\"" + value + "\"); instead of session.query().where(x=>x.Id == \"" + value + "\");\n" +
        "You can turn this error off by specifying documentStore.getConventions().setAllowQueriesOnId(true);, but that is not recommend and provided for backward compatibility reasons only.");
  }

  private void startTiming() {
    spStart = new Date().getTime();
  }

  public void logQuery() {
    log.debug("Executing query '%s' on index '%s' in '%s'",
        indexQuery.getQuery(), indexName, sessionOperations.getStoreIdentifier());
  }

  public AutoCloseable enterQueryContext() {
    if (firstRequest) {
      startTiming();
      firstRequest = false;
    }

    if (waitForNonStaleResults == false) {
      return null;
    }

    return sessionOperations.getDocumentStore().disableAggressiveCaching();
  }

  public boolean shouldQueryAgain(Exception e) {
    if (e instanceof NonAuthoritativeInformationException == false) {
      return false;
    }
    return (new Date().getTime()  - spStart) <= sessionOperations.getNonAuthoritativeInformationTimeout();
  }

  public <T> List<T> complete(Class<T> clazz)
  {
    QueryResult queryResult = currentQueryResults.createSnapshot();
    for (RavenJObject include : queryResult.getIncludes()) {
      RavenJObject metadata = include.value(RavenJObject.class, "@metadata");

      sessionOperations.trackEntity(Object.class, metadata.value(String.class, "@id"), include, metadata, disableEntitiesTracking);
    }

    List<T> list = new ArrayList<>();
    for (RavenJObject obj: queryResult.getResults()) {
      list.add(deserialize(clazz, obj));
    }

    sessionOperations.registerMissingIncludes(queryResult.getResults(), includes);

    return list;
  }

  public boolean isDisableEntitiesTracking() {
    return disableEntitiesTracking;
  }

  public void setDisableEntitiesTracking(boolean disableEntitiesTracking) {
    this.disableEntitiesTracking = disableEntitiesTracking;
  }

  @SuppressWarnings("unchecked")
  public <T> T deserialize(Class<T> clazz, RavenJObject result) {
    RavenJObject metadata = result.value(RavenJObject.class, "@metadata");
    if ((projectionFields == null || projectionFields.length <= 0) &&
        (metadata != null && StringUtils.isNotEmpty(metadata.value(String.class, "@id")))) {
      return (T) sessionOperations.trackEntity(clazz, metadata.value(String.class, "@id"), result, metadata, disableEntitiesTracking);
    }

    if (RavenJObject.class.equals(clazz)) {
      return (T)result;
    }

    String documentId = result.value(String.class, Constants.DOCUMENT_ID_FIELD_NAME); //check if the result contain the reserved name

    if (StringUtils.isNotEmpty(documentId) && String.class.equals(clazz) && // __document_id is present, and result type is a string
        projectionFields != null && projectionFields.length== 1 && // We are projecting one field only (although that could be derived from the
        // previous check, one could never be too careful
        ((metadata != null && result.getCount() == 2) || (metadata == null && result.getCount() == 1)) // there are no more props in the result object
        ) {
      return (T)documentId;
    }

    handleInternalMetadata(result);

    T deserializedResult = deserializedResult(clazz, result);

    if (StringUtils.isNotEmpty(documentId)) {
      // we need to make an additional check, since it is possible that a value was explicitly stated
      // for the identity property, in which case we don't want to override it.
      Field identityProperty = sessionOperations.getConventions().getIdentityProperty(clazz);
      if (identityProperty == null ||
          (!result.containsKey(identityProperty.getName()) ||
              result.get(identityProperty.getName()).getType() == JTokenType.NULL))  {
        sessionOperations.getGenerateEntityIdOnTheClient().trySetIdentity(deserializedResult, documentId);
      }
    }

    return deserializedResult;
  }

  private <T> T deserializedResult(Class<T> clazz, RavenJObject result) {
    if (projectionFields != null && projectionFields.length == 1) { // we only select a single field
      if (String.class.equals(clazz) || Number.class.isAssignableFrom(clazz) || clazz.isEnum()) {
        return result.value(clazz, projectionFields[0]);
      }
    }

    try {
      return JsonExtensions.createDefaultJsonSerializer().readValue(result.toString(), clazz);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private void handleInternalMetadata(RavenJObject result) {
    // Implant a property with "id" value ... if not exists
    RavenJObject metadata = result.value(RavenJObject.class, "@metadata");
    if (metadata == null || StringUtils.isEmpty(metadata.value(String.class, "@id"))) {
      // if the item has metadata, then nested items will not have it, so we can skip recursing down

      for (Entry<String, RavenJToken> nestedTuple : result) {
        RavenJToken nested = nestedTuple.getValue();
        if (nested instanceof RavenJObject) {
          handleInternalMetadata((RavenJObject) nested);
        }
        if (nested instanceof RavenJArray) {
          RavenJArray array = (RavenJArray) nested;
          for (RavenJToken token : array) {
            if (token instanceof RavenJObject) {
              handleInternalMetadata((RavenJObject) token);
            }
          }
        }
      }
      return;

    }

    String entityName = metadata.value(String.class, Constants.RAVEN_ENTITY_NAME);

    String idPropName = sessionOperations.getConventions().getFindIdentityPropertyNameFromEntityName().find(entityName);

    if (result.containsKey(idPropName)) {
      return;
    }
    result.add(idPropName, new RavenJValue(metadata.value(String.class, "@id")));
  }

  public void forceResult(QueryResult result)
  {
    currentQueryResults = result;
    currentQueryResults.ensureSnapshot();
  }

  public boolean isAcceptable(QueryResult result) {
    if (!sessionOperations.isAllowNonAuthoritativeInformation() &&
        result.isNonAuthoritativeInformation()) {
      if ((new Date().getTime() - spStart) > sessionOperations.getNonAuthoritativeInformationTimeout()) {
        spStop = new Date().getTime();
        throw new RuntimeException( String.format("Waited for %dms for the query to return authoritative result.",
            (spStop - spStart)));
      }
      log.debug(
          "Non authoritative query results on authoritative query '%s' on index '%s' in '%s', query will be retried, index etag is: %s",
          indexQuery.getQuery(),
          indexName,
          sessionOperations.getStoreIdentifier(),
          result.getIndexEtag());
      return false;
    }
    if (waitForNonStaleResults && result.isStale()) {
      if ((new Date().getTime() - spStart) > timeout) {
        spStop = new Date().getTime();
        throw new RuntimeException(
            String.format("Waited for %sms for the query to return non stale result.",
                (spStop - spStart)));
      }
      log.debug(
          "Stale query results on non stale query '%s' on index '%s' in '%s', query will be retried, index etag is: %s",
          indexQuery.getQuery(),
          indexName,
          sessionOperations.getStoreIdentifier(),
          result.getIndexEtag());
      return false;
    }
    currentQueryResults = result;
    currentQueryResults.ensureSnapshot();
    log.debug("Query returned %d/%d %sresults", result.getResults().size(),
        result.getTotalResults(),(result.isStale() ? "stale " : ""));
    return true;
  }

  private void addOperationHeaders() {
    if (setOperationHeaders == null)
      return;

    for (Tuple<String, Class<?>> sortByHint : sortByHints) {
      if (sortByHint.getItem2() == null) {
        continue;
      }
      setOperationHeaders.apply(String.format("SortHint-%s", UrlUtils.escapeDataString(StringUtils.strip(sortByHint.getItem1(), "-"))),
          SharpEnum.value(sessionOperations.getConventions().getDefaultSortOption(sortByHint.getClass().getCanonicalName())));
    }

  }

}
